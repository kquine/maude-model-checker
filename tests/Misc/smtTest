#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/smtTest.maude -no-banner -no-advise \
  > smtTest.out 2>&1

diff $srcdir/smtTest.expected smtTest.out > /dev/null 2>&1
