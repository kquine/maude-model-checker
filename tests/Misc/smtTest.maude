set show timing off .

load smt

*** Base theory
fmod TEST-B is
  pr BOOLEAN .
  vars W X Y Z : Boolean .
endfm

check W != X and Y .
check W = X and Y .
check X != tt and X != Y and Y != tt .
check X != tt and X != Y and Y != ff .
check (X xor tt) and (X xor Y) and (Y xor tt) .
check (X xor tt) and (X xor Y) and (Y xor ff) .
check not (X or not X) .
check not(X implies Y implies X) .
check (X ? Y : Z) and (Y implies X) and (Z implies X) .
check (X ? Y : Z) and (Y implies not X) and (Z implies X) .
check X = Y and Y = Z .
check X = Y and Y = Z and Z != X .

*** Theory of integers
fmod TEST-I is
  pr INTEGER .
  vars W X Y Z : Boolean .
  vars I J K L : Integer .
endfm

check I > J and J > K .
check I > J and J > K and K > I .
check I < J and J < K .
check I < J and J < K and K < I .
check I >= J and J >= I .
check I >= J and J >= I and I != J .
check I <= J and J <= I .
check I <= J and J <= I and I != J .

check I = J + K and I > J and I > K .
check I = J + K and I > J + K .
check I = J + K and I > J + K + L .
check I = J - K and I > J + K .
check - I = J - K and I > J + K .
check I != J + K and I > J + K .
check (I > J ? I : J) >= I and (I > J ? I : J) >= J .
check (I > J ? I : J) < I .
check (I > J ? I : J) = I or (I > J ? I : J) = J .
check not((I > J ? I : J) = I or (I > J ? I : J) = J) .
check (I > J ? I : J) != I and (I > J ? I : J) != J .

*** Theory of reals
fmod TEST-R is
  pr REAL .
  vars W X Y Z : Boolean .
  vars P Q R S : Real .
endfm

check P > Q and Q > R .
check P > Q and Q > R and R > P .
check P < Q and Q < R .
check P < Q and Q < R and R < P .
check P >= Q and Q >= P .
check P >= Q and Q >= P and P != Q .
check P <= Q and Q <= P .
check P <= Q and Q <= P and P != Q .

check P = Q + R and P > Q and P > R .
check P = Q + R and P > Q + R .
check P = Q + R and P > Q + R + s .
check P = Q - R and P > Q + R .
check - P = Q - R and P > Q + R .
check P != Q + R and P > Q + R .
check (P > Q ? P : Q) >= P and (P > Q ? P : Q) >= Q .
check (P > Q ? P : Q) < P .
check (P > Q ? P : Q) = P or (P > Q ? P : Q) = Q .
check not((P > Q ? P : Q) = P or (P > Q ? P : Q) = Q) .
check (P > Q ? P : Q) != P and (P > Q ? P : Q) != Q .

*** Theory of reals + integers
fmod TEST-RI is
  pr REAL-INTEGER .
  vars W X Y Z : Boolean .
  vars I J K L : Integer .
  vars P Q R S : Real .
endfm

check isInteger(toReal(I) + toReal(J)) != ff .
check isInteger(toReal(I) + toReal(J)) != tt .
check toInteger(R) + toInteger(P) = toInteger(R + P) .
check not(toInteger(R) + toInteger(P) = toInteger(R + P)) .

check 1/2 < P .
check 1/2 < P and P < 3/4 .
check 1/2 < P and P < 3/4 and isInteger(P) .
check -2 < I .
check -2 < I and -2 * I > -1 .
check -2 < I and -2 * I > -1 and I != -1 .
check -2 < I and -2 * I > -1 and I != -1 and I - I != I .

*** meta-level version
select META-LEVEL .

red metaCheck(['REAL-INTEGER], '_=_['_+_['1.Integer, '1.Integer], '2.Integer]) .

red metaCheck(['REAL-INTEGER], '_=_['_+_['1.Integer, '1.Integer], '3.Integer]) .

fmod META-CHECK is
  pr META-LEVEL .
  pr REAL-INTEGER .
  vars W X Y Z : Boolean .
  vars I J K L : Integer .
  vars P Q R S : Real .
endfm

red metaCheck(['META-CHECK], upTerm((I > J ? I : J) < I)) .
red metaCheck(['META-CHECK], upTerm((I > J ? I : J) > I)) .

red metaCheck(['META-CHECK], upTerm( (I > J ? I : J) = I or (I > J ? I : J) = J  )) .
red metaCheck(['META-CHECK], upTerm( not((I > J ? I : J) = I or (I > J ? I : J) = J ) )) .
red metaCheck(['META-CHECK], upTerm( (I > J ? I : J) != I and (I > J ? I : J) != J )) .
red metaCheck(['META-CHECK], upTerm( (I > J ? I : J) != I or (I > J ? I : J) != J )) .
