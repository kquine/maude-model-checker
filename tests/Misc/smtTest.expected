Maude> ==========================================
check in TEST-B : W != X and Y .
Result from CVC4 is: sat
==========================================
check in TEST-B : W = X and Y .
Result from CVC4 is: sat
==========================================
check in TEST-B : X != tt and X != Y and Y != tt .
Result from CVC4 is: unsat
==========================================
check in TEST-B : X != tt and X != Y and Y != ff .
Result from CVC4 is: sat
==========================================
check in TEST-B : (X xor tt) and (X xor Y) and (Y xor tt) .
Result from CVC4 is: unsat
==========================================
check in TEST-B : (X xor tt) and (X xor Y) and (Y xor ff) .
Result from CVC4 is: sat
==========================================
check in TEST-B : not (X or not X) .
Result from CVC4 is: unsat
==========================================
check in TEST-B : not (X implies Y implies X) .
Result from CVC4 is: unsat
==========================================
check in TEST-B : (X ? Y : Z) and (Y implies X) and (Z implies X) .
Result from CVC4 is: sat
==========================================
check in TEST-B : (X ? Y : Z) and (Y implies not X) and (Z implies X) .
Result from CVC4 is: unsat
==========================================
check in TEST-B : X = Y and Y = Z .
Result from CVC4 is: sat
==========================================
check in TEST-B : X = Y and Y = Z and Z != X .
Result from CVC4 is: unsat
==========================================
check in TEST-I : I > J and J > K .
Result from CVC4 is: sat
==========================================
check in TEST-I : I > J and J > K and K > I .
Result from CVC4 is: unsat
==========================================
check in TEST-I : I < J and J < K .
Result from CVC4 is: sat
==========================================
check in TEST-I : I < J and J < K and K < I .
Result from CVC4 is: unsat
==========================================
check in TEST-I : I >= J and J >= I .
Result from CVC4 is: sat
==========================================
check in TEST-I : I >= J and J >= I and I != J .
Result from CVC4 is: unsat
==========================================
check in TEST-I : I <= J and J <= I .
Result from CVC4 is: sat
==========================================
check in TEST-I : I <= J and J <= I and I != J .
Result from CVC4 is: unsat
==========================================
check in TEST-I : I = J + K and I > J and I > K .
Result from CVC4 is: sat
==========================================
check in TEST-I : I = J + K and I > J + K .
Result from CVC4 is: unsat
==========================================
check in TEST-I : I = J + K and I > J + K + L .
Result from CVC4 is: sat
==========================================
check in TEST-I : I = J - K and I > J + K .
Result from CVC4 is: sat
==========================================
check in TEST-I : - I = J - K and I > J + K .
Result from CVC4 is: sat
==========================================
check in TEST-I : I != J + K and I > J + K .
Result from CVC4 is: sat
==========================================
check in TEST-I : (I > J ? I : J) >= I and (I > J ? I : J) >= J .
Result from CVC4 is: sat
==========================================
check in TEST-I : (I > J ? I : J) < I .
Result from CVC4 is: unsat
==========================================
check in TEST-I : (I > J ? I : J) = I or (I > J ? I : J) = J .
Result from CVC4 is: sat
==========================================
check in TEST-I : not ((I > J ? I : J) = I or (I > J ? I : J) = J) .
Result from CVC4 is: unsat
==========================================
check in TEST-I : (I > J ? I : J) != I and (I > J ? I : J) != J .
Result from CVC4 is: unsat
==========================================
check in TEST-R : P > Q and Q > R .
Result from CVC4 is: sat
==========================================
check in TEST-R : P > Q and Q > R and R > P .
Result from CVC4 is: unsat
==========================================
check in TEST-R : P < Q and Q < R .
Result from CVC4 is: sat
==========================================
check in TEST-R : P < Q and Q < R and R < P .
Result from CVC4 is: unsat
==========================================
check in TEST-R : P >= Q and Q >= P .
Result from CVC4 is: sat
==========================================
check in TEST-R : P >= Q and Q >= P and P != Q .
Result from CVC4 is: unsat
==========================================
check in TEST-R : P <= Q and Q <= P .
Result from CVC4 is: sat
==========================================
check in TEST-R : P <= Q and Q <= P and P != Q .
Result from CVC4 is: unsat
==========================================
check in TEST-R : P = Q + R and P > Q and P > R .
Result from CVC4 is: sat
==========================================
check in TEST-R : P = Q + R and P > Q + R .
Result from CVC4 is: unsat
Warning: <standard input>, line 70: bad token s.
Warning: <standard input>, line 70: no parse for term.
==========================================
check in TEST-R : P = Q - R and P > Q + R .
Result from CVC4 is: sat
==========================================
check in TEST-R : - P = Q - R and P > Q + R .
Result from CVC4 is: sat
==========================================
check in TEST-R : P != Q + R and P > Q + R .
Result from CVC4 is: sat
==========================================
check in TEST-R : (P > Q ? P : Q) >= P and (P > Q ? P : Q) >= Q .
Result from CVC4 is: sat
==========================================
check in TEST-R : (P > Q ? P : Q) < P .
Result from CVC4 is: unsat
==========================================
check in TEST-R : (P > Q ? P : Q) = P or (P > Q ? P : Q) = Q .
Result from CVC4 is: sat
==========================================
check in TEST-R : not ((P > Q ? P : Q) = P or (P > Q ? P : Q) = Q) .
Result from CVC4 is: unsat
==========================================
check in TEST-R : (P > Q ? P : Q) != P and (P > Q ? P : Q) != Q .
Result from CVC4 is: unsat
==========================================
check in TEST-RI : isInteger(toReal(I) + toReal(J)) != ff .
Result from CVC4 is: sat
==========================================
check in TEST-RI : isInteger(toReal(I) + toReal(J)) != tt .
Result from CVC4 is: unsat
==========================================
check in TEST-RI : toInteger(R) + toInteger(P) = toInteger(R + P) .
Result from CVC4 is: sat
==========================================
check in TEST-RI : not toInteger(R) + toInteger(P) = toInteger(R + P) .
Result from CVC4 is: sat
==========================================
check in TEST-RI : 1/2 < P .
Result from CVC4 is: sat
==========================================
check in TEST-RI : 1/2 < P and P < 3/4 .
Result from CVC4 is: sat
==========================================
check in TEST-RI : 1/2 < P and P < 3/4 and isInteger(P) .
Result from CVC4 is: unsat
==========================================
check in TEST-RI : -2 < I .
Result from CVC4 is: sat
==========================================
check in TEST-RI : -2 < I and -2 * I > -1 .
Result from CVC4 is: sat
==========================================
check in TEST-RI : -2 < I and -2 * I > -1 and I != -1 .
Result from CVC4 is: sat
==========================================
check in TEST-RI : -2 < I and -2 * I > -1 and I != -1 and I - I != I .
Result from CVC4 is: unsat
==========================================
reduce in META-LEVEL : metaCheck(['REAL-INTEGER], '_=_['_+_['1.Integer,
    '1.Integer],'2.Integer]) .
rewrites: 2
result Bool: true
==========================================
reduce in META-LEVEL : metaCheck(['REAL-INTEGER], '_=_['_+_['1.Integer,
    '1.Integer],'3.Integer]) .
rewrites: 2
result Bool: false
==========================================
reduce in META-CHECK : metaCheck(['META-CHECK], upTerm((I > J ? I : J) < I)) .
rewrites: 3
result Bool: false
==========================================
reduce in META-CHECK : metaCheck(['META-CHECK], upTerm((I > J ? I : J) > I)) .
rewrites: 3
result Bool: true
==========================================
reduce in META-CHECK : metaCheck(['META-CHECK], upTerm((I > J ? I : J) = I or (
    I > J ? I : J) = J)) .
rewrites: 3
result Bool: true
==========================================
reduce in META-CHECK : metaCheck(['META-CHECK], upTerm(not ((I > J ? I : J) = I
    or (I > J ? I : J) = J))) .
rewrites: 3
result Bool: false
==========================================
reduce in META-CHECK : metaCheck(['META-CHECK], upTerm((I > J ? I : J) != I and
    (I > J ? I : J) != J)) .
rewrites: 3
result Bool: false
==========================================
reduce in META-CHECK : metaCheck(['META-CHECK], upTerm((I > J ? I : J) != I or
    (I > J ? I : J) != J)) .
rewrites: 3
result Bool: true
Maude> Bye.
